Use NativePromise for WebCodecs encode/decode/flush methods

youennf opened the pull request:
#### 39a1102365fe8338a6ee5082b47000ae8a1560e9
<pre>
Use NativePromise for WebCodecs encode/decode/flush methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=278909">https://bugs.webkit.org/show_bug.cgi?id=278909</a>
<a href="https://rdar.apple.com/135005637">rdar://135005637</a>

Reviewed by Philippe Normand and Jean-Yves Avenard.

We move all these methods to NativePromise as it is easier to read and is easier for maintenance.
For instance, the NativePromises are wrapped via ScriptExecutionContext::enqueueTaskWhenSettled so that
it is guaranteed for the callback to be executed and destroyed in the context thread.
This will allow us to use ActiveDOMObject pending activities as part of these callbacks in a follow-up.
Another step will be to migrate the create encoder/decoder methods to NativePromise as well.

* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::WebCodecsAudioDecoder::decode):
(WebCore::WebCodecsAudioDecoder::flush):
(WebCore::WebCore::WebCodecsAudioDecoder::suspend):
(WebCore::WebCodecsAudioDecoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
(WebCore::WebCodecsAudioEncoder::configure):
(WebCore::WebCodecsAudioEncoder::encode):
(WebCore::WebCodecsAudioEncoder::flush):
(WebCore::WebCore::WebCodecsAudioEncoder::suspend):
(WebCore::WebCodecsAudioEncoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::decode):
(WebCore::WebCodecsVideoDecoder::flush):
(WebCore::WebCore::WebCodecsVideoDecoder::suspend):
(WebCore::WebCodecsVideoDecoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::WebCodecsVideoEncoder::updateRates):
(WebCore::WebCodecsVideoEncoder::configure):
(WebCore::WebCodecsVideoEncoder::encode):
(WebCore::WebCodecsVideoEncoder::flush):
(WebCore::WebCore::WebCodecsVideoEncoder::suspend):
(WebCore::WebCodecsVideoEncoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.h:
* Source/WebCore/platform/AudioDecoder.h:
* Source/WebCore/platform/AudioEncoder.h:
* Source/WebCore/platform/VideoDecoder.h:
* Source/WebCore/platform/VideoEncoder.h:
(WebCore::VideoEncoder::setRates): Deleted.
* Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
(WebCore::GStreamerAudioDecoder::decode):
(WebCore::GStreamerAudioDecoder::flush):
(WebCore::GStreamerInternalAudioDecoder::GStreamerInternalAudioDecoder):
(WebCore::GStreamerInternalAudioDecoder::decode):
(WebCore::GStreamerInternalAudioDecoder::flush):
* Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.h:
* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
(WebCore::GStreamerAudioEncoder::encode):
(WebCore::GStreamerAudioEncoder::flush):
(WebCore::GStreamerInternalAudioEncoder::GStreamerInternalAudioEncoder):
(WebCore::GStreamerInternalAudioEncoder::flush):
* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSample):
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerVideoDecoder::decode):
(WebCore::GStreamerVideoDecoder::flush):
(WebCore::GStreamerInternalVideoDecoder::decode):
(WebCore::GStreamerInternalVideoDecoder::flush):
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::create):
(WebCore::GStreamerVideoEncoder::create):
(WebCore::GStreamerVideoEncoder::GStreamerVideoEncoder):
(WebCore::GStreamerVideoEncoder::encode):
(WebCore::GStreamerVideoEncoder::flush):
(WebCore::GStreamerVideoEncoder::setRates):
(WebCore::GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder):
(WebCore::GStreamerInternalVideoEncoder::initialize):
(WebCore::GStreamerInternalVideoEncoder::setRates):
(WebCore::GStreamerInternalVideoEncoder::applyRates):
(WebCore::GStreamerInternalVideoEncoder::flush):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.h:
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
(WebCore::LibWebRTCVPXVideoDecoder::decode):
(WebCore::LibWebRTCVPXVideoDecoder::flush):
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.h:
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.cpp:
(WebCore::LibWebRTCVPXVideoEncoder::encode):
(WebCore::LibWebRTCVPXVideoEncoder::flush):
(WebCore::LibWebRTCVPXVideoEncoder::setRates):
(WebCore::LibWebRTCVPXInternalVideoEncoder::encode):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.h:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.messages.in:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::flushDecoder):
(WebKit::LibWebRTCCodecsProxy::setEncodeRates):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoDecoder::decode):
(WebKit::RemoteVideoDecoder::flush):
(WebKit::RemoteVideoEncoder::encode):
(WebKit::RemoteVideoEncoder::setRates):
(WebKit::RemoteVideoEncoder::flush):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::decodeVideoFrame):
(WebKit::LibWebRTCCodecs::flushDecoder):
(WebKit::LibWebRTCCodecs::sendFrameToDecode):
(WebKit::LibWebRTCCodecs::decodeWebRTCFrame):
(WebKit::LibWebRTCCodecs::decodeFrame):
(WebKit::LibWebRTCCodecs::decodeFrameInternal):
(WebKit::LibWebRTCCodecs::encodeFrameInternal):
(WebKit::LibWebRTCCodecs::encodeFrame):
(WebKit::LibWebRTCCodecs::flushEncoder):
(WebKit::LibWebRTCCodecs::setEncodeRates):
(WebKit::LibWebRTCCodecs::gpuProcessConnectionDidClose):
(WebKit::LibWebRTCCodecs::setDecoderConnection):
(WebKit::LibWebRTCCodecs::flushDecoderCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.messages.in:

Canonical link: <a href="https://commits.webkit.org/283145@main">https://commits.webkit.org/283145@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f14058e7073d8c51cfc7ed05852a284a552f65fe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65227 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44596 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69251 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15833 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67345 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16115 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52405 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10964 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41221 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56461 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33028 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37892 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13832 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14710 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59776 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14172 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70956 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13650 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59728 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9211 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56522 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60003 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14407 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7587 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1260 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40406 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41483 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42664 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41227 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -302,13 +309,14 @@ void WebCodecsAudioDecoder::processControlMessageQueue()
 
 void WebCore::WebCodecsAudioDecoder::suspend(ReasonForSuspension)
 {
-    // FIXME: Implement.
```
seems unrelated

youennf replied:
All our async tasks are wrapped by enqueuTask so I think this is now fine and we can remove the FIXME

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -417,12 +420,14 @@ void WebCodecsAudioEncoder::processControlMessageQueue()
 }
 
 void WebCodecsAudioEncoder::stop()
 {
```
Can't we disconnect the pending NativePromises?

youennf replied:
Ditto here.

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -177,11 +180,15 @@ ExceptionOr<void> WebCodecsAudioDecoder::flush(Ref<DeferredPromise>&& promise)
     queueControlMessageAndProcess({ *this, [this, clearFlushPromiseCount = m_clearFlushPromiseCount] {
-        m_internalDecoder->flush([this, clearFlushPromiseCount, pendingActivity = takePendingWebCodecActivity()] {
-            if (clearFlushPromiseCount != m_clearFlushPromiseCount)
+        protectedScriptExecutionContext()->enqueueTaskWhenSettled(m_internalDecoder->flush(), TaskSource::MediaElement, [weakThis = ThreadSafeWeakPtr { *this }, clearFlushPromiseCount, pendingActivity = takePendingWebCodecActivity()] (auto&&) {
```
we could probably get rid of `takePendingWebCodecActivity()`, tracking if we have any pending promises instead.

youennf replied:
The plan is to go back to makePendingActivity. But I'll do this in a follow-up.

jyavenard commented on Source/WebCore/platform/AudioEncoder.h:
```diff
@@ -34,7 +34,7 @@
 #include <span>
 #include <wtf/CompletionHandler.h>
-#include <wtf/Expected.h>
+#include <wtf/NativePromise.h>
```
`wtf/Forward.h` should be sufficient here

jyavenard commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -334,16 +335,14 @@ void GStreamerInternalAudioDecoder::decode(std::span<const uint8_t> frameData, b
-
-        callback({ });
     });
+    return AudioDecoder::DecodePromise::resolve();
```
this is incorrect, we are returning a resolved promise immediately before the work as completed.
We have to create a NativePromise::Producer pass it to the lambda and resolve it where `callback` was called.

youennf replied:
This is inside GStreamerInternalAudioDecoder not GStreamerAudioDecoder  so the work has completed.

jyavenard commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -290,22 +292,21 @@ GStreamerInternalAudioDecoder::GStreamerInternalAudioDecoder(const String& codec
-            callback({ });
         });
-        return;
+        return AudioDecoder::DecodePromise::reject("Empty frame"_s);
```
That will be changing the order of event, promise should be rejected in the lambda, in place of the callback()

youennf replied:
My hope is to remove entirely postTask.
I don't think this is changing the order of events:
1. postTask is doing: hop to context thread and enqueue a task to execute the callback,
2. resolve will hop to context thread in enqueueTaskWhenSettled via nativePromiseDispatcher, and then will enqueue a task to execute the callback.

jyavenard commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -56,7 +56,7 @@ class GStreamerInternalAudioEncoder : public ThreadSafeRefCountedAndCanMakeThrea
     void postTask(Function<void()>&& task) { m_postTaskCallback(WTFMove(task)); }
     bool encode(AudioEncoder::RawFrame&&);
-    void flush(Function<void()>&&);
+    void flush();
```
we now have two flush() methods, one returning a Ref<GenericPromise> and the other returning void.

youennf replied:
Yeah will fix this.

jyavenard commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -139,34 +139,28 @@ GStreamerAudioEncoder::~GStreamerAudioEncoder()
-
-        String resultString;
-        if (result)
-            encoder->harness()->processOutputSamples();
```
this part is no longer required? seems incorrect.

youennf replied:
Right, I need to add it back.

jyavenard commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -139,34 +139,28 @@ GStreamerAudioEncoder::~GStreamerAudioEncoder()
-    gstEncoderWorkQueue().dispatch([encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
-        encoder->flush(WTFMove(callback));
+    return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder] {
+        encoder->flush();
```
could make the encoder return a promise too and chain them together

youennf replied:
We would do that if flush was async, but it is not.

jyavenard replied:
it was async before. could have kept it async and do :
```
return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder] {
        return encoder->flush();
});
```

jyavenard commented on Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
```diff
@@ -251,22 +251,21 @@ GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder(const String& codec
-            callback({ });
         });
-        return;
+        return VideoDecoder::DecodePromise::createAndReject("Empty frame"_s);;
```
same as earlier comment.

jyavenard commented on Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
```diff
@@ -292,21 +291,18 @@ void GStreamerInternalVideoDecoder::decode(std::span<const uint8_t> frameData, b
 }
 
-void GStreamerInternalVideoDecoder::flush(Function<void()>&& callback)
+void GStreamerInternalVideoDecoder::flush()
```
that's different to the definition of the header, should return a Promise

jyavenard commented on Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
```diff
@@ -127,33 +127,25 @@ GStreamerVideoEncoder::~GStreamerVideoEncoder()
-
-        String resultString;
-        if (result)
-            encoder->harness()->processOutputSamples();
```
no more call to processOutputSamples() ?

jyavenard commented on Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
```diff
@@ -398,17 +398,15 @@ int32_t LibWebRTCCodecs::releaseDecoder(Decoder& decoder)
-    decoder.connection->send(Messages::LibWebRTCCodecsProxy::FlushDecoder { decoder.identifier }, 0);
-    Locker flushLocker { decoder.flushCallbacksLock };
-    decoder.flushCallbacks.append(WTFMove(callback));
+    return decoder.connection->sendWithPromisedReply(Messages::LibWebRTCCodecsProxy::FlushDecoder { decoder.identifier }, 0)->whenSettled(workQueue(), [] (auto&&) {
```
I think we should be able to directly return the promise with the right PromiseConverter, removing the need to dispatch another task.

youennf replied:
I have that in another patch, and will do in a follow-up as it requires some changes to Connection.h.
I want to remove all these whenSettled at once to remove some potential race conditions between decode/flush.

jyavenard commented on Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
```diff
@@ -421,30 +419,48 @@ void LibWebRTCCodecs::setDecoderFormatDescription(Decoder& decoder, std::span<co
 
-    decoder.connection->sendWithPromisedReply(Messages::LibWebRTCCodecsProxy::DecodeFrame { decoder.identifier, timeStamp, data }, 0)->whenSettled(workQueue(), [callback = WTFMove(callback)] (auto&& result) mutable {
-        callback(result ? result.value() : false);
+    decoder.connection->sendWithPromisedReply(Messages::LibWebRTCCodecsProxy::DecodeFrame { decoder.identifier, timeStamp, data }, 0)->whenSettled(workQueue(), [producer = WTFMove(producer)] (auto&& result) mutable {
```
same here, we don't need the optional Producer, we can have sendFrameToDecode return a promise too and directly return the sendWithPromisedReply value with the PromiseConverter template

youennf replied:
Ditto.

philn commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -261,6 +262,7 @@ GStreamerInternalAudioDecoder::GStreamerInternalAudioDecoder(const String& codec
 
     m_harness = GStreamerElementHarness::create(WTFMove(harnessedElement), [weakThis = WeakPtr { *this }, this](auto&, GRefPtr<GstSample>&& outputSample) {
+        RefPtr protectedThis = weakThis.get();
         if (!weakThis)
```
Should we check `protectedThis` here now?

youennf replied:
Ah right, that would be better!

youennf replied:
Done

philn commented on Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
```diff
@@ -127,33 +130,35 @@ GStreamerVideoEncoder::~GStreamerVideoEncoder()
-        encoder->flush(WTFMove(callback));
+    return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder, bitRate, frameRate] {
+        encoder->setRates(bitRate, frameRate);
+        encoder->applyRates();
```
`applyRates()` is called by `setRates()` already, so maybe no need to call it a second time here?

youennf replied:
Done

jyavenard reviewed (approved):
r=me for changes other than those with GTK, that we resolve the promise immediately rather than in the `m_postTaskCallback` needs to be reviewed by someone more familiar with the code,

jyavenard commented on Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
```diff
@@ -334,16 +335,14 @@ void GStreamerInternalAudioDecoder::decode(std::span<const uint8_t> frameData, b
-
-        callback({ });
     });
+    return AudioDecoder::DecodePromise::createAndResolve();
```
here I'm not 100% on this, we are resolving the promise concurrently to calling `processOutputSamples()` this is a different code flow than before.

Could we resolve the promise there instead?

philn replied:
Hm yes, this might be racy now. I'm taking a look right now, please don't merge this yet.

philn replied:
@youennf can you apply this please? It makes the audio decoder behavior consistent with the video decoder.

```diff
diff --git a/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp b/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp
index a34f180817ab..aa91f200c333 100644
--- a/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp
+++ b/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp
@@ -325,15 +325,15 @@ Ref<AudioDecoder::DecodePromise> GStreamerInternalAudioDecoder::decode(std::span
         GST_BUFFER_DURATION(encodedData.get()) = *duration;
 
     auto result = m_harness->pushSample(adoptGRef(gst_sample_new(encodedData.get(), m_inputCaps.get(), &segment, nullptr)));
+    if (result)
+        m_harness->processOutputSamples();
     m_postTaskCallback([weakThis = WeakPtr { *this }, this, result]() mutable {
         if (!weakThis)
             return;
         if (m_isClosed)
             return;
 
-        if (result)
-            m_harness->processOutputSamples();
-        else
+        if (!result)
             m_outputCallback(makeUnexpected("Decode error"_s));
     });
     return AudioDecoder::DecodePromise::createAndResolve();
```

youennf replied:
This raciness would be observable with GC issues.
There is no race if `m_harness->processOutputSamples()` calls synchronously m_outputCallback.
That said, I think this should not be more racy with this patch than before.

philn replied:
The `processOutputSamples()` will synchronously call the callback passed to the `GStreamerElementHarness::create()` function.

philn replied:
so `m_outputCallback` ends-up being called, but from `m_postTaskCallback()`.

youennf replied:
Yep, I'll do the follow-up of removing `m_outputCallback` today or tomorrow.
This will indeed make the code clearer.

youennf replied:
I'll make the gstreamer audio change, thanks for checking it.
I am surprised this audio decoding issue was not hit by GTK/WPE bots on WPT audio tests.

philn replied:
TBH I haven't noticed a change in tests behavior with this change and without. At least when running the tests locally. Let's see how the bots behave after landing this!

jyavenard replied:
@philn  with your change we should call m_postTaskCallback only if `!result` is true ? so that there's no need to capture `result`

philn replied:
Right... but anyway iiuc the postTaskCallback stuff is going away soon?

youennf committed:
Use NativePromise for WebCodecs encode/decode/flush methods
https://bugs.webkit.org/show_bug.cgi?id=278909
rdar://135005637

Reviewed by Philippe Normand and Jean-Yves Avenard.

We move all these methods to NativePromise as it is easier to read and is easier for maintenance.
For instance, the NativePromises are wrapped via ScriptExecutionContext::enqueueTaskWhenSettled so that
it is guaranteed for the callback to be executed and destroyed in the context thread.
This will allow us to use ActiveDOMObject pending activities as part of these callbacks in a follow-up.
Another step will be to migrate the create encoder/decoder methods to NativePromise as well.

* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::WebCodecsAudioDecoder::decode):
(WebCore::WebCodecsAudioDecoder::flush):
(WebCore::WebCore::WebCodecsAudioDecoder::suspend):
(WebCore::WebCodecsAudioDecoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
(WebCore::WebCodecsAudioEncoder::configure):
(WebCore::WebCodecsAudioEncoder::encode):
(WebCore::WebCodecsAudioEncoder::flush):
(WebCore::WebCore::WebCodecsAudioEncoder::suspend):
(WebCore::WebCodecsAudioEncoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::decode):
(WebCore::WebCodecsVideoDecoder::flush):
(WebCore::WebCore::WebCodecsVideoDecoder::suspend):
(WebCore::WebCodecsVideoDecoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::WebCodecsVideoEncoder::updateRates):
(WebCore::WebCodecsVideoEncoder::configure):
(WebCore::WebCodecsVideoEncoder::encode):
(WebCore::WebCodecsVideoEncoder::flush):
(WebCore::WebCore::WebCodecsVideoEncoder::suspend):
(WebCore::WebCodecsVideoEncoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.h:
* Source/WebCore/platform/AudioDecoder.h:
* Source/WebCore/platform/AudioEncoder.h:
* Source/WebCore/platform/VideoDecoder.h:
* Source/WebCore/platform/VideoEncoder.h:
(WebCore::VideoEncoder::setRates): Deleted.
* Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp:
(WebCore::GStreamerAudioDecoder::decode):
(WebCore::GStreamerAudioDecoder::flush):
(WebCore::GStreamerInternalAudioDecoder::GStreamerInternalAudioDecoder):
(WebCore::GStreamerInternalAudioDecoder::decode):
(WebCore::GStreamerInternalAudioDecoder::flush):
* Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.h:
* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
(WebCore::GStreamerAudioEncoder::encode):
(WebCore::GStreamerAudioEncoder::flush):
(WebCore::GStreamerInternalAudioEncoder::GStreamerInternalAudioEncoder):
(WebCore::GStreamerInternalAudioEncoder::flush):
* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSample):
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerVideoDecoder::decode):
(WebCore::GStreamerVideoDecoder::flush):
(WebCore::GStreamerInternalVideoDecoder::decode):
(WebCore::GStreamerInternalVideoDecoder::flush):
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::create):
(WebCore::GStreamerVideoEncoder::create):
(WebCore::GStreamerVideoEncoder::GStreamerVideoEncoder):
(WebCore::GStreamerVideoEncoder::encode):
(WebCore::GStreamerVideoEncoder::flush):
(WebCore::GStreamerVideoEncoder::setRates):
(WebCore::GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder):
(WebCore::GStreamerInternalVideoEncoder::initialize):
(WebCore::GStreamerInternalVideoEncoder::setRates):
(WebCore::GStreamerInternalVideoEncoder::applyRates):
(WebCore::GStreamerInternalVideoEncoder::flush):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.h:
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
(WebCore::LibWebRTCVPXVideoDecoder::decode):
(WebCore::LibWebRTCVPXVideoDecoder::flush):
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.h:
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.cpp:
(WebCore::LibWebRTCVPXVideoEncoder::encode):
(WebCore::LibWebRTCVPXVideoEncoder::flush):
(WebCore::LibWebRTCVPXVideoEncoder::setRates):
(WebCore::LibWebRTCVPXInternalVideoEncoder::encode):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoEncoder.h:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.h:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.messages.in:
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::flushDecoder):
(WebKit::LibWebRTCCodecsProxy::setEncodeRates):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoDecoder::decode):
(WebKit::RemoteVideoDecoder::flush):
(WebKit::RemoteVideoEncoder::encode):
(WebKit::RemoteVideoEncoder::setRates):
(WebKit::RemoteVideoEncoder::flush):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::decodeVideoFrame):
(WebKit::LibWebRTCCodecs::flushDecoder):
(WebKit::LibWebRTCCodecs::sendFrameToDecode):
(WebKit::LibWebRTCCodecs::decodeWebRTCFrame):
(WebKit::LibWebRTCCodecs::decodeFrame):
(WebKit::LibWebRTCCodecs::decodeFrameInternal):
(WebKit::LibWebRTCCodecs::encodeFrameInternal):
(WebKit::LibWebRTCCodecs::encodeFrame):
(WebKit::LibWebRTCCodecs::flushEncoder):
(WebKit::LibWebRTCCodecs::setEncodeRates):
(WebKit::LibWebRTCCodecs::gpuProcessConnectionDidClose):
(WebKit::LibWebRTCCodecs::setDecoderConnection):
(WebKit::LibWebRTCCodecs::flushDecoderCompleted): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.messages.in:

Canonical link: https://commits.webkit.org/283145@main
